storage/replica_command.go

```diff
@@ -250,6 +250,10 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
 	if args.Txn == nil {
 		return reply, nil, util.Errorf("no transaction specified to EndTransaction")
 	}
+	if !bytes.Equal(args.Key, args.Txn.Key) {
```

Comment:
I might not be understanding code correctly, but I think we need this to avoid processing commands (e.g., `HeartbeatTxn` and `EndTransaction`) for the same txn at the same time. The command queue looks at `args.Key` to block a command whose keys overlap with already-running commands.